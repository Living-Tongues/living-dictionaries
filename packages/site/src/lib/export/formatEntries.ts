import type {
  IDictionary,
  IEntry,
  IPartOfSpeech,
  ISemanticDomain,
  ISpeaker,
} from '@living-dictionaries/types';
import { glossingLanguages } from './glossing-languages-temp'; // todo - import from actual json file
import { friendlyName } from './friendlyName';
import { stripHTMLTags } from './stripHTMLTags';

enum EntryCSVFieldsEnum {
  id = 'Entry Id',
  lx = 'Lexeme/Word/Phrase',
  ph = 'Phonetic (IPA)',
  in = 'Interlinearization',
  nc = 'Noun class',
  mr = 'Morphology',
  pl = 'Plural form',
  di = 'Dialect',
  nt = 'Notes',
  sr = 'Source(s)',
  psab = 'Part of Speech abbreviation',
  ps = 'Part of Speech',
  sfFriendlyName = 'Audio filename',
  sfsn = 'Speaker name',
  sfbp = 'Speaker birthplace',
  sfde = 'Speaker decade',
  sfge = 'Speaker gender',
  pfFriendlyName = 'Image filename',
}
type EntryForCSVKeys = keyof typeof EntryCSVFieldsEnum;
type EntryForCSV = {
  [key in EntryForCSVKeys]?: string;
};
export interface IEntryForCSV extends EntryForCSV {
  xsvn?: string;
  va?: string; // optional for Babanki
  sfpa?: string; // for downloading file, not exported in CSV
  pfpa?: string; // for downloading file, not exported in CSV
}

export function formatEntriesForCSV(
  entries: IEntry[],
  { name: dictionaryName, id: dictionaryId, glossLanguages }: IDictionary,
  speakers: ISpeaker[],
  semanticDomains: ISemanticDomain[],
  partsOfSpeech: IPartOfSpeech[]
) {
  const headers = {} as IEntryForCSV;
  for (const key in EntryCSVFieldsEnum) {
    headers[key] = EntryCSVFieldsEnum[key];
  }

  // Begin dynamic headers

  // Assign max number of semantic domains used by a single entry
  const maxSDN = Math.max(...entries.map((entry) => entry.sdn?.length || 0));
  if (maxSDN > 0) {
    for (let index = 0; index < maxSDN; index++) {
      headers[`sd${index + 1}`] = `Semantic domain ${index + 1}`;
    }
  }

  // glosses
  glossLanguages.forEach((bcp) => {
    headers[`gl${bcp}`] = `${glossingLanguages[bcp]} Gloss`;
  });

  // Vernacular and gloss language example sentence headers
  headers.xsvn = `Example sentence in ${dictionaryName}`;
  glossLanguages.forEach((bcp) => {
    headers[`xs${bcp}`] = `Example sentence in ${glossingLanguages[bcp]}`;
  });

  // Dictionary specific
  if (dictionaryId === 'babanki') {
    headers.va = 'variant';
  }

  const formattedEntries: IEntryForCSV[] = entries.map((entry) => {
    const formattedEntry = {
      id: entry.id,
      lx: entry.lx,
      ph: entry.ph,
      in: entry.in,
      nc: entry.nc,
      mr: entry.mr,
      pl: entry.pl,
      di: entry.di,
      nt: stripHTMLTags(entry.nt),
      sr: entry.sr ? typeof entry.sr === 'string' ? entry.sr : entry.sr.join(' | ') : '', // some dictionaries (e.g. Kalanga) have sources that are strings and not arrays
      psab: '',
      ps: '',
      sfFriendlyName: '',
      sfsn: '',
      sfbp: '',
      sfde: '',
      sfge: '',
      pfFriendlyName: '',
    } as IEntryForCSV;

    // part of speech (abbreviation & name)
    if (entry.ps) {
      const fullPos = partsOfSpeech.find((ps) => ps.enAbbrev === entry.ps)?.enName;
      if (!fullPos) {
        formattedEntry.ps = entry.ps;
      } else {
        formattedEntry.psab = entry.ps;
        formattedEntry.ps = fullPos;
      }
    }

    // Media
    if (entry.sf?.path) {
      const speaker = speakers.find((speaker) => speaker?.id === entry.sf.sp);
      const speakerName = speaker?.displayName || entry.sf.speakerName;
      formattedEntry.sfpa = entry.sf.path;
      formattedEntry.sfFriendlyName = friendlyName(entry, entry.sf.path);
      formattedEntry.sfsn = speakerName;
      formattedEntry.sfbp = speaker?.birthplace;
      formattedEntry.sfde = speaker?.decade?.toString();
      formattedEntry.sfge = speaker?.gender;
    }
    if (entry.pf?.path) {
      formattedEntry.pfpa = entry.pf.path;
      formattedEntry.pfFriendlyName = friendlyName(entry, entry.pf.path);
    }

    // Begin dynamic values

    // semantic domains
    for (let index = 0; index < maxSDN; index++) {
      formattedEntry[`sd${index + 1}`] = '';
      if (entry.sdn && entry.sdn[index]) {
        const matchingDomain = semanticDomains.find((sd) => sd.key === entry.sdn[index]);
        if (matchingDomain) {
          formattedEntry[`sd${index + 1}`] = matchingDomain.name;
        }
      }
    }

    // glosses
    glossLanguages.forEach((bcp) => {
      const cleanEntry = stripHTMLTags(entry.gl[bcp]);
      formattedEntry[`gl${bcp}`] = cleanEntry;
    });

    // Vernacular and gloss language example sentences
    formattedEntry.xsvn = entry.xs?.vn;
    glossLanguages.forEach((bcp) => {
      formattedEntry[`xs${bcp}`] = entry.xs?.[bcp];
    });

    // Dictionary specific
    if (dictionaryId === 'babanki') {
      formattedEntry.va = entry.va;
    }

    return formattedEntry;
  });
  return [headers, ...formattedEntries];
}
