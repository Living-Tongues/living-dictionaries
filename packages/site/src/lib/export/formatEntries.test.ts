import type { IDictionary, IEntry, ISemanticDomain, ISpeaker } from '@living-dictionaries/types';
import { formatEntriesForCSV } from './formatEntries';

test('formatEntriesForCSV basic example to smoke test', () => {
  const entriesArray: IEntry[] = [
    {
      id: '12ar',
      lx: 'banana',
      lo: 'کیلا',
      ph: 'banana',
      gl: { es: 'platano' },
      ps: '1',
      sdn: ['2.1', '2.2', '2.3'],
      di: 'dialect x',
      nt: 'This is an example of a note, here we can write whatever we want.',
      sr: ['A book', 'www.mybook.com'],
      xs: { en: 'This is a banana' },
      sf: { path: 'https://example.com', sp: 'aa34' },
      pf: { gcs: 'https://example.com', path: 'https://example.com' },
    },
    { id: '34qw', lx: 'tree', gl: { es: 'arbol' } },
  ];
  const dictionary: IDictionary = { name: 'test', glossLanguages: ['ar', 'en'], entryCount: 0 };
  const speakers: ISpeaker[] = [{ displayName: 'John Smith', id: '123' }];
  const semanticDomains: ISemanticDomain[] = [{ key: '2.1', name: 'Plant Test Domain' }];
  expect(formatEntriesForCSV(entriesArray, dictionary, speakers, semanticDomains, [])).toMatchSnapshot();
});
