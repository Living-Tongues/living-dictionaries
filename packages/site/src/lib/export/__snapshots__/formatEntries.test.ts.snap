// Vitest Snapshot v1

exports[`formatEntriesForCSV basic example to smoke test 1`] = `
[
  {
    "di": "Dialect",
    "glar": "Arabic Gloss",
    "glen": "English Gloss",
    "id": "Entry Id",
    "in": "Interlinearization",
    "lx": "Lexeme/Word/Phrase",
    "mr": "Morphology",
    "nc": "Noun class",
    "nt": "Notes",
    "pfFriendlyName": "Image filename",
    "ph": "Phonetic (IPA)",
    "ps": "Part of Speech",
    "psab": "Part of Speech abbreviation",
    "sd1": "Semantic domain 1",
    "sd2": "Semantic domain 2",
    "sd3": "Semantic domain 3",
    "sfFriendlyName": "Audio filename",
    "sfbp": "Speaker birthplace",
    "sfde": "Speaker decade",
    "sfge": "Speaker gender",
    "sfsn": "Speaker name",
    "sr": "Source(s)",
    "xsar": "Example sentence in Arabic",
    "xsen": "Example sentence in English",
    "xsvn": "Example sentence in test",
  },
  {
    "di": "dialect x",
    "glar": "",
    "glen": "",
    "id": "12ar",
    "in": "",
    "lx": "banana",
    "mr": "",
    "nc": "",
    "nt": "This is an example of a note - here we can write whatever we want.",
    "pfFriendlyName": "12ar_platano.com",
    "pfpa": "https://example.com",
    "ph": "banana",
    "pl": "",
    "ps": "1",
    "psab": "",
    "sd1": "Plant Test Domain",
    "sd2": "",
    "sd3": "",
    "sfFriendlyName": "12ar_platano.com",
    "sfbp": "",
    "sfde": "",
    "sfge": "",
    "sfpa": "https://example.com",
    "sfsn": "",
    "sr": "A book | www.mybook.com",
    "xsar": "",
    "xsen": "This is a banana",
    "xsvn": "",
  },
  {
    "di": "",
    "glar": "",
    "glen": "",
    "id": "34qw",
    "in": "",
    "lx": "tree",
    "mr": "",
    "nc": "",
    "nt": "",
    "pfFriendlyName": "",
    "ph": "",
    "pl": "",
    "ps": "",
    "psab": "",
    "sd1": "",
    "sd2": "",
    "sd3": "",
    "sfFriendlyName": "",
    "sfbp": "",
    "sfde": "",
    "sfge": "",
    "sfsn": "",
    "sr": "",
    "xsar": "",
    "xsen": "",
    "xsvn": "",
  },
]
`;
